Sources/StreamChat/Workers/ReadStateSender.swift

```diff
@@ -4,74 +4,96 @@
-            self.messageRepository = messageRepository
+final class ReadStateSender {
+    private let authenticationRepository: AuthenticationRepository
+    private let channelUpdater: ChannelUpdater
+    private let messageRepository: MessageRepository
+    @Atomic private var markingRead = false
+    @Atomic private(set) var isMarkedAsUnread = false
+    
+    init(
+        authenticationRepository: AuthenticationRepository,
+        channelUpdater: ChannelUpdater,
+        messageRepository: MessageRepository
+    ) {
+        self.authenticationRepository = authenticationRepository
+        self.channelUpdater = channelUpdater
+        self.messageRepository = messageRepository
+    }
+    
+    func markRead(_ channel: ChatChannel, completion: @escaping (Error?) -> Void) {
+        guard
+            !markingRead,
+            let currentUserId = authenticationRepository.currentUserId,
+            let currentUserRead = channel.reads.first(where: { $0.user.id == currentUserId }),
+            let lastMessageAt = channel.lastMessageAt,
+            currentUserRead.lastReadAt < lastMessageAt
+        else {
+            completion(nil)
+            return
+        }
+        markingRead = true
+        channelUpdater.markRead(cid: channel.cid, userId: currentUserId) { error in
+            self.markingRead = false
+            self.isMarkedAsUnread = false
+            completion(error)
         }
-        
-        private var markingReadTask: Task<Void, Error>?
-        private var markingUnreadTask: Task<Void, Error>?
-        
-        func markRead(_ channel: ChatChannel) async throws {
-            // Wait for marking unread
-            if let task = markingUnreadTask {
-                _ = try? await task.value
+    }
+    
+    @available(iOS 13.0, *)
+    func markRead(_ channel: ChatChannel) async throws {
+        try await withCheckedThrowingContinuation { continuation in
+            markRead(channel) { error in
+                continuation.resume(with: error)
             }
-            
-            if let active = markingReadTask {
-                try await active.value
-            } else {
-                guard let currentUserId = authenticationRepository.currentUserId else { throw ClientError.CurrentUserDoesNotExist() }
-                guard let lastMessageAt = channel.lastMessageAt else { return }
-                guard let userRead = channel.reads.first(where: { $0.user.id == currentUserId }) else { return }
-                guard userRead.lastReadAt < lastMessageAt else { return }
-                
-                markingReadTask = Task {
-                    defer { markingReadTask = nil }
-                    try await channelUpdater.markRead(cid: cid, userId: currentUserId)
+        }
+    }
+    
+    func markUnread(
+        from messageId: MessageId,
+        in channel: ChatChannel,
+        completion: @escaping (Result<ChatChannel, Error>) -> Void
+    ) {
+        guard !markingRead,
```

Comment:
I guess it's ok we use the same bool for marking as unread, do you see any issues with it? (seems like it was like that before too)

Reply:
I think it is important we use the same bool. We don't want to mark **unread**, while we are marking **read**, otherwise we will have race conditions. The problem is that it seems that we maybe lose some requests here. If it happens a user marks unread/read while one of the requests is being executed we don't do anything

Reply:
But I would say that is not a big deal at the moment

Reply:
Currently I just replicated what controller was doing and did not want to change any logic there. But that said, it does not seem like 100% reliable implementation when many unread and read calls happen.